File: backend/src/models/Chunk.js
import { query, queryOne, insert, update } from '../config/database.js';

/**
 * Create a new chunk
 */
export async function createChunk(data) {
  const id = await insert(
    `INSERT INTO chunks (
      document_id, chunk_index, chunk_hash, content, content_length,
      token_estimate, hierarchy_path, hierarchy_json, hierarchy_level,
      paragraph_start, paragraph_end, metadata
    ) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)`,
    [
      data.documentId,
      data.chunkIndex,
      data.chunkHash,
      data.content,
      data.contentLength,
      data.tokenEstimate,
      data.hierarchyPath,
      JSON.stringify(data.hierarchyJson || []),
      data.hierarchyLevel,
      data.paragraphStart,
      data.paragraphEnd,
      JSON.stringify(data.metadata || {}),
    ]
  );
  return id;
}

/**
 * Create multiple chunks in batch
 */
export async function createChunks(chunks) {
  const ids = [];
  for (const chunk of chunks) {
    const id = await createChunk(chunk);
    ids.push(id);
  }
  return ids;
}

/**
 * Get chunk by ID
 */
export async function getChunkById(id) {
  return queryOne('SELECT * FROM chunks WHERE id = ?', [id]);
}

/**
 * Get chunks by document ID
 */
export async function getChunksByDocumentId(documentId, options = {}) {
  const { limit = 1000, offset = 0 } = options;

  return query(
    `SELECT * FROM chunks WHERE document_id = ?
     ORDER BY chunk_index ASC LIMIT ${parseInt(limit)} OFFSET ${parseInt(offset)}`,
    [documentId]
  );
}

/**
 * Get all chunks with pagination
 */
export async function getAllChunks(options = {}) {
  const { documentId, limit = 100, offset = 0 } = options;

  let sql = `
    SELECT c.*, d.original_name as document_name
    FROM chunks c
    JOIN documents d ON c.document_id = d.id
  `;
  const params = [];

  if (documentId) {
    sql += ' WHERE c.document_id = ?';
    params.push(documentId);
  }

  sql += ` ORDER BY c.document_id, c.chunk_index LIMIT ${parseInt(limit)} OFFSET ${parseInt(offset)}`;

  return query(sql, params);
}

/**
 * Update chunk vector information
 */
export async function updateChunkVector(id, vectorId, embeddingModel) {
  return update(
    `UPDATE chunks SET vector_id = ?, embedded_at = NOW(), embedding_model = ? WHERE id = ?`,
    [vectorId, embeddingModel, id]
  );
}

/**
 * Delete chunks by document ID
 */
export async function deleteChunksByDocumentId(documentId) {
  return update('DELETE FROM chunks WHERE document_id = ?', [documentId]);
}

/**
 * Get chunk count by document
 */
export async function getChunkCountByDocument(documentId) {
  const result = await queryOne(
    'SELECT COUNT(*) as count FROM chunks WHERE document_id = ?',
    [documentId]
  );
  return result?.count || 0;
}

/**
 * Get chunk statistics
 */
export async function getChunkStats() {
  const stats = await queryOne(`
    SELECT
      COUNT(*) as total_chunks,
      AVG(token_estimate) as avg_tokens,
      MIN(token_estimate) as min_tokens,
      MAX(token_estimate) as max_tokens,
      AVG(content_length) as avg_length,
      COUNT(DISTINCT document_id) as document_count
    FROM chunks
  `);

  return {
    totalChunks: stats?.total_chunks || 0,
    avgTokens: Math.round(stats?.avg_tokens || 0),
    minTokens: stats?.min_tokens || 0,
    maxTokens: stats?.max_tokens || 0,
    avgLength: Math.round(stats?.avg_length || 0),
    documentCount: stats?.document_count || 0,
  };
}

/**
 * Get chunks by IDs
 */
export async function getChunksByIds(ids) {
  if (!ids || ids.length === 0) return [];

  const placeholders = ids.map(() => '?').join(',');
  return query(
    `SELECT c.*, d.original_name as document_name
     FROM chunks c
     JOIN documents d ON c.document_id = d.id
     WHERE c.id IN (${placeholders})`,
    ids
  );
}

/**
 * Get hierarchical tree structure for a document
 */
export async function getChunkTree(documentId) {
  const chunks = await query(
    `SELECT id, chunk_index, hierarchy_path, hierarchy_json, hierarchy_level,
            content, token_estimate, content_length
     FROM chunks WHERE document_id = ? ORDER BY chunk_index`,
    [documentId]
  );

  // Build tree structure
  const tree = { children: [], chunks: [] };

  for (const chunk of chunks) {
    const path = JSON.parse(chunk.hierarchy_json || '[]');
    let current = tree;

    for (let i = 0; i < path.length; i++) {
      const name = path[i];
      let child = current.children.find(c => c.name === name);

      if (!child) {
        child = { name, level: i + 1, children: [], chunks: [] };
        current.children.push(child);
      }

      current = child;
    }

    current.chunks.push({
      id: chunk.id,
      index: chunk.chunk_index,
      tokenEstimate: chunk.token_estimate,
      contentLength: chunk.content_length,
      preview: chunk.content.substring(0, 150) + (chunk.content.length > 150 ? '...' : ''),
    });
  }

  return tree;
}

/**
 * Get hierarchy summary for a document (for debugging)
 */
export async function getHierarchySummary(documentId) {
  const chunks = await query(
    `SELECT id, chunk_index, hierarchy_path, hierarchy_json, hierarchy_level,
            token_estimate, content_length, SUBSTRING(content, 1, 100) as preview
     FROM chunks WHERE document_id = ? ORDER BY chunk_index`,
    [documentId]
  );

  const withHierarchy = [];
  const withoutHierarchy = [];

  for (const chunk of chunks) {
    const hierarchyJson = JSON.parse(chunk.hierarchy_json || '[]');
    const summary = {
      id: chunk.id,
      index: chunk.chunk_index,
      hierarchyPath: chunk.hierarchy_path || '(none)',
      hierarchyLevel: chunk.hierarchy_level || 0,
      tokens: chunk.token_estimate,
      length: chunk.content_length,
      preview: chunk.preview + '...',
    };

    if (hierarchyJson.length > 0) {
      withHierarchy.push(summary);
    } else {
      withoutHierarchy.push(summary);
    }
  }

  return {
    totalChunks: chunks.length,
    chunksWithHierarchy: withHierarchy.length,
    chunksWithoutHierarchy: withoutHierarchy.length,
    withHierarchy,
    withoutHierarchy,
  };
}

/**
 * Search chunks by text content
 */
export async function searchChunksText(searchText, documentId = null, limit = 50) {
  let sql = `
    SELECT c.*, d.original_name as document_name
    FROM chunks c
    JOIN documents d ON c.document_id = d.id
    WHERE c.content LIKE ?
  `;
  const params = [`%${searchText}%`];

  if (documentId) {
    sql += ' AND c.document_id = ?';
    params.push(documentId);
  }

  sql += ` ORDER BY c.document_id, c.chunk_index LIMIT ${parseInt(limit)}`;

  return query(sql, params);
}

File: backend-python/models/chunk.py
"""
Chunk model - CRUD operations for document chunks
"""
import json
from database import get_cursor
from typing import Optional, List, Dict, Any


def create_chunk(document_id: int, chunk_index: int, content: str,
                 content_length: int, token_estimate: int,
                 chunk_hash: Optional[str] = None,
                 hierarchy_path: Optional[str] = None,
                 hierarchy_json: Optional[List] = None,
                 hierarchy_level: Optional[int] = None,
                 paragraph_start: Optional[int] = None,
                 paragraph_end: Optional[int] = None) -> int:
    """Create a new chunk"""
    with get_cursor() as cursor:
        cursor.execute("""
            INSERT INTO chunks
            (document_id, chunk_index, content, content_length, token_estimate,
             content_hash, hierarchy_path, hierarchy_json, hierarchy_level,
             paragraph_start, paragraph_end)
            VALUES (%s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s)
        """, (
            document_id, chunk_index, content, content_length, token_estimate,
            chunk_hash, hierarchy_path,
            json.dumps(hierarchy_json) if hierarchy_json else None,
            hierarchy_level, paragraph_start, paragraph_end
        ))
        return cursor.lastrowid


def get_chunk_by_id(chunk_id: int) -> Optional[Dict[str, Any]]:
    """Get chunk by ID with document info"""
    with get_cursor() as cursor:
        cursor.execute("""
            SELECT c.*, d.original_name as document_name
            FROM chunks c
            JOIN documents d ON c.document_id = d.id
            WHERE c.id = %s
        """, (chunk_id,))
        row = cursor.fetchone()
        if row and row.get('hierarchy_json'):
            try:
                row['hierarchy_json'] = json.loads(row['hierarchy_json'])
            except:
                pass
        return row


def get_chunks(document_id: Optional[int] = None, limit: int = 50,
               offset: int = 0) -> List[Dict[str, Any]]:
    """Get chunks with optional document filter"""
    with get_cursor() as cursor:
        if document_id:
            cursor.execute("""
                SELECT c.id, c.document_id, c.content, c.content_length,
                       c.token_estimate, c.hierarchy_path, c.hierarchy_level,
                       c.vector_id, d.original_name as document_name
                FROM chunks c
                JOIN documents d ON c.document_id = d.id
                WHERE c.document_id = %s
                ORDER BY c.id
                LIMIT %s OFFSET %s
            """, (document_id, limit, offset))
        else:
            cursor.execute("""
                SELECT c.id, c.document_id, c.content, c.content_length,
                       c.token_estimate, c.hierarchy_path, c.hierarchy_level,
                       c.vector_id, d.original_name as document_name
                FROM chunks c
                JOIN documents d ON c.document_id = d.id
                ORDER BY c.id DESC
                LIMIT %s OFFSET %s
            """, (limit, offset))
        return cursor.fetchall()


def get_chunk_count_by_document(document_id: int) -> int:
    """Get chunk count for a document"""
    with get_cursor() as cursor:
        cursor.execute(
            "SELECT COUNT(*) as count FROM chunks WHERE document_id = %s",
            (document_id,)
        )
        result = cursor.fetchone()
        return result['count'] if result else 0


def get_chunk_stats() -> Dict[str, Any]:
    """Get chunk statistics"""
    with get_cursor() as cursor:
        cursor.execute("""
            SELECT
                COUNT(*) as total_chunks,
                ROUND(AVG(token_estimate)) as avg_tokens,
                MIN(token_estimate) as min_tokens,
                MAX(token_estimate) as max_tokens
            FROM chunks
        """)
        result = cursor.fetchone()
        return {
            'totalChunks': result['total_chunks'] or 0,
            'avgTokens': result['avg_tokens'] or 0,
            'minTokens': result['min_tokens'] or 0,
            'maxTokens': result['max_tokens'] or 0
        }


def delete_chunks_by_document_id(document_id: int):
    """Delete all chunks for a document"""
    with get_cursor() as cursor:
        cursor.execute("DELETE FROM chunks WHERE document_id = %s", (document_id,))


def update_chunk_vector(chunk_id: int, vector_id: str, embedding_model: str):
    """Update chunk with vector ID"""
    with get_cursor() as cursor:
        cursor.execute("""
            UPDATE chunks SET vector_id = %s, embedding_model = %s
            WHERE id = %s
        """, (str(vector_id), embedding_model, chunk_id))


def search_chunks_text(query: str, document_id: Optional[int] = None,
                       limit: int = 50) -> List[Dict[str, Any]]:
    """Full-text search in chunks"""
    with get_cursor() as cursor:
        search_pattern = f"%{query}%"
        if document_id:
            cursor.execute("""
                SELECT c.id, c.document_id, c.content, c.content_length,
                       c.token_estimate, c.hierarchy_path, c.hierarchy_level,
                       c.vector_id, d.original_name as document_name
                FROM chunks c
                JOIN documents d ON c.document_id = d.id
                WHERE c.document_id = %s AND c.content LIKE %s
                ORDER BY c.id
                LIMIT %s
            """, (document_id, search_pattern, limit))
        else:
            cursor.execute("""
                SELECT c.id, c.document_id, c.content, c.content_length,
                       c.token_estimate, c.hierarchy_path, c.hierarchy_level,
                       c.vector_id, d.original_name as document_name
                FROM chunks c
                JOIN documents d ON c.document_id = d.id
                WHERE c.content LIKE %s
                ORDER BY c.id
                LIMIT %s
            """, (search_pattern, limit))
        return cursor.fetchall()


def get_chunks_by_ids(chunk_ids: List[int]) -> List[Dict[str, Any]]:
    """Get multiple chunks by IDs"""
    if not chunk_ids:
        return []
    with get_cursor() as cursor:
        placeholders = ','.join(['%s'] * len(chunk_ids))
        cursor.execute(f"""
            SELECT c.*, d.original_name as document_name
            FROM chunks c
            JOIN documents d ON c.document_id = d.id
            WHERE c.id IN ({placeholders})
        """, tuple(chunk_ids))
        return cursor.fetchall()


def get_chunk_tree(document_id: int) -> Dict[str, Any]:
    """Build hierarchical tree structure for document chunks"""
    with get_cursor() as cursor:
        cursor.execute("""
            SELECT id, content, hierarchy_path, hierarchy_json,
                   hierarchy_level, token_estimate
            FROM chunks
            WHERE document_id = %s
            ORDER BY id
        """, (document_id,))
        chunks = cursor.fetchall()

    # Build tree structure
    tree = {'name': 'root', 'children': [], 'chunks': []}

    for chunk in chunks:
        hierarchy_json = chunk.get('hierarchy_json')
        if hierarchy_json:
            try:
                hierarchy = json.loads(hierarchy_json) if isinstance(hierarchy_json, str) else hierarchy_json
            except:
                hierarchy = []
        else:
            hierarchy = []

        chunk_data = {
            'id': chunk['id'],
            'preview': chunk['content'][:100] + '...' if len(chunk['content']) > 100 else chunk['content'],
            'tokenEstimate': chunk['token_estimate']
        }

        if not hierarchy:
            # Orphan chunk (no hierarchy)
            tree['chunks'].append(chunk_data)
        else:
            # Navigate/create tree path
            current = tree
            for i, node_info in enumerate(hierarchy):
                node_name = node_info.get('name', 'Unknown')
                node_level = node_info.get('level', i + 1)

                # Find or create child node
                child = next(
                    (c for c in current.get('children', []) if c['name'] == node_name),
                    None
                )
                if not child:
                    child = {
                        'name': node_name,
                        'level': node_level,
                        'children': [],
                        'chunks': []
                    }
                    if 'children' not in current:
                        current['children'] = []
                    current['children'].append(child)
                current = child

            # Add chunk to leaf node
            current['chunks'].append(chunk_data)

    return tree


def get_hierarchy_summary(document_id: int) -> Dict[str, int]:
    """Get hierarchy summary for a document"""
    with get_cursor() as cursor:
        cursor.execute("""
            SELECT
                COUNT(*) as total_chunks,
                SUM(CASE WHEN hierarchy_path IS NOT NULL AND hierarchy_path != '' THEN 1 ELSE 0 END) as with_hierarchy,
                SUM(CASE WHEN hierarchy_path IS NULL OR hierarchy_path = '' THEN 1 ELSE 0 END) as without_hierarchy
            FROM chunks
            WHERE document_id = %s
        """, (document_id,))
        result = cursor.fetchone()
        return {
            'totalChunks': result['total_chunks'] or 0,
            'chunksWithHierarchy': result['with_hierarchy'] or 0,
            'chunksWithoutHierarchy': result['without_hierarchy'] or 0
        }
